### ISSUE TITLE
Poor arm64 / embedded performance; huge memory footprint

### ISSUE THREAD

**Issue by elitak** [ipfs/go-ipfs#4370]
#### Version information:
go-ipfs version: 0.4.11-
Repo version: 6
System version: arm64/linux
Golang version: go1.8.5

#### Description:
I'm trying to run the ipfs daemon on a cheap, small, low-power arm boards, and it frequently consumes so much memory that the whole system pretty much grinds to a halt. This is usually during operations that involve large files with many chunks, e.g.`ipfs pin add`ing a large number of hashes present on the local network.

These systems have 4x 1GHz cores and 512 or 1024M of DDR3 RAM. Adding in swap space doesn't seem to help.

Has memory profiling been done on the daemon? Is there a way to limit or reduce the memory footprint of the process at the cost of maybe slower I/O? Incidentally, I/O is pretty slow as it is, only achieving about 1MB/s during an `ipfs get` over my LAN, whereas rsync between the two hosts exceeds 30MB/s. Is that expected even for x86_64 hosts?

**Comment by Stebalien** [comment]
> Has memory profiling been done on the daemon? Is there a way to limit or reduce the memory footprint of the process at the cost of maybe slower I/O?

The next release (I'm building RC2 now) should help significantly. Much of our memory usage comes from keeping many connections open and the next release includes logic to close the less important ones.

Another issue you're likely hitting, in terms of poor bandwidth utilization and memory usage, is providing. Currently, every time you download a block, you tell the network that you have the block. This is expensive and often includes opening multiple connections to different peers. At the cost of *not* telling the network what blocks you have (peers looking for that block won't be able to find it unless they already have a connection to you) you can add files with the `--local` flag. We have an in-progress feature that will avoid announcing *every* block to the network that should help significantly but that won't make it into the next release.

**Comment by elitak** [comment]
I should add that this is happening between 2 hosts that only have each other as peers. I've added everything except my local network to `Swarm.AddrFilters`, and I'm only seeing a single connection between them in `netstat`, so I don't think connections are the issue.

**Comment by Kubuxu** [comment]
The fact that we are using AES will have also important implications here. 98% x86 have hardware acceleration for AES.

See the comparison:
```
On i7-6800k:
BenchmarkEncryptAesCbc-12           500000          3541 ns/op     406.57 MB/s
BenchmarkDescryptAesCbc-12          500000          3007 ns/op     478.76 MB/s
BenchmarkXSalsa20-12               1000000          1670 ns/op     854.37 MB/s

On RPi 3:
BenchmarkEncryptAesCbc-4           10000        203769 ns/op       7.07 MB/s
BenchmarkDescryptAesCbc-4          10000        202064 ns/op       7.13 MB/s
BenchmarkXSalsa20-4                10000        107729 ns/op      13.25 MB/s
```

**Comment by Stebalien** [comment]
How much memory *are* they using? I'd expect ~200MiB.

**Comment by elitak** [comment]
This is the device I'm using:
http://wiki.friendlyarm.com/wiki/index.php/NanoPi_NEO_Plus2

Here's 1 of the 4 cpu entries in /proc/cpuinfo:
processor       : 0
BogoMIPS        : 48.00
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x0
CPU part        : 0xd03
CPU revision    : 4

It has some kind of acceleration, I guess. @Kubuxu What do I use to get the same benchmark?

**Comment by Kubuxu** [comment]
https://github.com/Kubuxu/go-crypto-bench/ You have to install go toolkit and run `go test -bench` in that directory.

**Comment by Stebalien** [comment]
This may also be due to hashing. sha256 has been optimized on x86_64 (amd64) but not on arm. We should consider using https://github.com/minio/sha256-simd for arm.

**Comment by Stebalien** [comment]
@elitak Would you be willing to test out a custom IPFS build?

**Comment by elitak** [comment]
@Stebalien Yes, definitely, to the extent where I can build it from a given commit and will run it and report whether I see the same issues.

Currently, the ipfs daemon just DoSes the board shortly after boot. The RAM gets so overcommitted it just grinds to a halt after like 10-30m. This is with the daemon doing nothing but connecting to 3 other peers on my LAN. I don't know if the daemon queues pins and resumes them between runs, but even if it does, I don't think that's what's burdening them. Why does the daemon need like 4GB of RAM while completely idle?

**Comment by Stebalien** [comment]
> Why does the daemon need like 4GB of RAM while completely idle?

It shouldn't, that's why I asked *how* much memory it was using. This is definitely *not* the problem I assumed.

> ipfs pin add

When you say you're adding a bunch of hashes, are you adding them individually or via a root hash. If you're adding them individually, it may be related to #4122 (although pins *shouldn't* resume when you restart your daemon). This has been fixed in master.

---

So, could you try:

1. Building the latest master (there have been a lot of fixes since 0.4.11). Note: You'll need go 1.9 to build it. If building it is going to be a pain, 0.4.13 should get you most of the important fixes (except the fix for #4122).
2. Following the [debugging guide](https://github.com/ipfs/go-ipfs/blob/master/docs/debug-guide.md#beginning) and posting the debug information somewhere (after you're daemon has been running for a while). This will allow us to figure out what exactly is going wrong.

**Comment by Stebalien** [comment]
Closing due to age.
